torch/csrc/autograd/record_function.cpp

```diff
@@ -26,39 +29,85 @@ void popCallback() {
   }
   start_callbacks.pop_back();
   end_callbacks.pop_back();
-  has_callbacks = !start_callbacks.empty();
+  needs_inputs_flags.pop_back();
+  if (needs_inputs_flags.size() > 1) {
+    if (needs_inputs_flags.back()) {
+      record_function_state = RecordFunctionState::NAME_AND_INPUTS;
+    } else {
+      record_function_state = RecordFunctionState::NAME_ONLY;
+    }
+  } else {
+    record_function_state = RecordFunctionState::NO_CALLBACKS;
+  }
+}
+
+TORCH_API RecordFunctionState recordFunctionState() {
+  return record_function_state;
+}
+
+RecordFunction::RecordFunction(Function* fn) {
+  before(fn);
```

Comment:
nit: `before` is a surprising name. Perhaps `init` would be better suited?

Reply:
it's not only to initialize, but also to call constructors _before_ the user code is run